vowpalwabbit/gd.cc

```diff
@@ -645,6 +649,12 @@ float compute_update(gd& g, example& ec)
   if (sparse_l2)
     update -= g.sparse_l2 * ec.pred.scalar;
 
+  if (std::isnan(update))
+    {
+      std::cout << "update is NAN, replacing with 0" << std::endl;
```

Comment:
this will ignore --quiet probably, not sure if on purpose. using all.trace_message << should respect it

Reply:
but don't actually understand when we use std::cerr vs std::cout since you also use std::cerr on line 533

Reply:
I shifted to cerr.    

The reporting issue is nontrivial here.  When we are deep inside an optimization loop and we discover some sort of fatal flaw, how should we report?   Not by throwing because that leads to bad states.  On the comandline I can print.  Used via library?   (What if tracemessage is unavailable?)

Reply:
Yeap, just pointing that out in case it was intentional to use cout, but cerr I understand.

Just a thought, but we could also augment 'example' to be able to be marked as faulty since that is the root cause of the failure. That way it could be communicated outside the hot path that the root cause of the failure comes from that specific example.

Reply:
Use the example to carry error info seems pretty reasonable to me.  It offers a nice synchronous path.